{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2020-05-13\n",
    "# Redoing all gene list creation, stringent and non-stringent\n",
    "# Want two lists, one ENS and one HGNC\n",
    "# convert each individual data stream from whatever source it is in to the other, flow that into the joint lists\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import sys\n",
    "import os\n",
    "import mygene\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_d = '/Users/cweiss19/Box Sync/MPRA/DATA/loci_to_genes/2020_05_13_bothlists'\n",
    "our_vars_file = '/Users/cweiss19/Box Sync/MPRA/VARs_anno.xlsx'\n",
    "\n",
    "# GeneHancer\n",
    "GH_connections = f'{p_d}/GeneHancer_gene_associations_scores.txt'\n",
    "GH_enhancers = f'{p_d}/GeneHancer_elements.txt'\n",
    "\n",
    "# HiC\n",
    "HiC = f'{p_d}/Jung_2019_Supp_table_3.xlsx'\n",
    "\n",
    "# eQTL and overlapping promoters in VARs_anno.xlsx\n",
    "\n",
    "# HGNC <-> ENSid\n",
    "ens_to_hgnc = f'{p_d}/ens--hgnc_conv_dict.pickle'\n",
    "hgnc_to_ens = f'{p_d}/hgnc--ens_conv_dict.pickle'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load conversion dicts\n",
    "with open(ens_to_hgnc, 'rb') as handle:\n",
    "    e_to_h_d = pickle.load(handle)  # {ensID: 'hgnc1,hgnc2', ensID2: 'hgnc3'} etc\n",
    "\n",
    "with open(hgnc_to_ens, 'rb') as handle:\n",
    "    h_to_e_d = pickle.load(handle)   # same as above but hgnc keys\n",
    "    \n",
    "    \n",
    "# all the ensIDs that are in one are in the other, same for hgnc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define ultimate function to convert genes from one to the other, applied to pandas DF rows\n",
    "\n",
    "def convert(r, which_key, column):\n",
    "    if which_key == 'ens':\n",
    "        d = e_to_h_d\n",
    "    if which_key == 'hgnc':\n",
    "        d = h_to_e_d\n",
    "    #print(r)\n",
    "    genes = r[column]  # will be single gene, or genes separated by ','\n",
    "    if not genes == 0:\n",
    "        genes_l = genes.split(',')\n",
    "        converted_genes_l = []\n",
    "        for g in genes_l:\n",
    "            if g in d:\n",
    "                conv_g = d[g]  # could be single gene, or genes separated by ','\n",
    "                conv_g_l = conv_g.split(',')  # turn into list\n",
    "                converted_genes_l.extend(conv_g_l)\n",
    "        converted_genes = ','.join(list(set(converted_genes_l)))\n",
    "        if len(converted_genes) > 0:\n",
    "            return converted_genes\n",
    "        else: \n",
    "            return 0\n",
    "    else:\n",
    "        return 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_vars = pd.read_excel(\n",
    "    our_vars_file, \n",
    "    index_col=0, \n",
    "    header=2, \n",
    "    sheet_name=0\n",
    "    )\n",
    "df_vars.drop(df_vars.index[0], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14042\n",
      "4797\n",
      "14042\n",
      "1922\n",
      "14042\n",
      "639\n",
      "14042\n",
      "1851\n",
      "                                                                          genes  \\\n",
      "oligo                                                                             \n",
      "libB_seq108668_[chrX:152673474-152673673]                     PNMA3,ZFP92,PNMA5   \n",
      "libA_seq46943_[chrX:153178234-153178433]   NAA10,RENBP,ARHGAP4,IDH3G,SRPK3,SSR4   \n",
      "\n",
      "                                                  esc_genes npc_genes  \\\n",
      "oligo                                                                   \n",
      "libB_seq108668_[chrX:152673474-152673673]             ZFP92     ZFP92   \n",
      "libA_seq46943_[chrX:153178234-153178433]   IDH3G,SRPK3,SSR4         0   \n",
      "\n",
      "                                          msc_genes  \\\n",
      "oligo                                                 \n",
      "libB_seq108668_[chrX:152673474-152673673]     ZFP92   \n",
      "libA_seq46943_[chrX:153178234-153178433]          0   \n",
      "\n",
      "                                                                                   genes_ens  \\\n",
      "oligo                                                                                          \n",
      "libB_seq108668_[chrX:152673474-152673673]    ENSG00000183837,ENSG00000189420,ENSG00000198883   \n",
      "libA_seq46943_[chrX:153178234-153178433]   ENSG00000180879,ENSG00000102030,ENSG0000010203...   \n",
      "\n",
      "                                                                             esc_genes_ens  \\\n",
      "oligo                                                                                        \n",
      "libB_seq108668_[chrX:152673474-152673673]                                  ENSG00000189420   \n",
      "libA_seq46943_[chrX:153178234-153178433]   ENSG00000184343,ENSG00000067829,ENSG00000180879   \n",
      "\n",
      "                                             msc_genes_ens    npc_genes_ens  \n",
      "oligo                                                                        \n",
      "libB_seq108668_[chrX:152673474-152673673]  ENSG00000189420  ENSG00000189420  \n",
      "libA_seq46943_[chrX:153178234-153178433]                 0                0  \n"
     ]
    }
   ],
   "source": [
    "# HiC overlap\n",
    "\n",
    "dist_from_var_HiC = 10000\n",
    "\n",
    "def read_in_excel(filepath):\n",
    "    df = pd.read_excel(filepath,\n",
    "        index_col=None,\n",
    "        header=0,\n",
    "        sheet_name=0)\n",
    "    return df\n",
    "\n",
    "def rows_to_dict(df, dict1):\n",
    "    for index, row in df.iterrows():\n",
    "        pos = row['Interacting_fragment']\n",
    "        gene = row['Promoter']\n",
    "        if pos in dict1:\n",
    "            dict1[pos].append(str(gene))\n",
    "        else:\n",
    "            dict1[pos] = [str(gene)]\n",
    "    return dict1\n",
    "\n",
    "def create_range_dict(dict1, dict2):\n",
    "    keys = list(dict1.keys())\n",
    "    for each_key in keys:\n",
    "        split_pos = each_key.split('.')\n",
    "        chromosome = split_pos[0]\n",
    "        if chromosome in dict2:\n",
    "            dict2[chromosome].append(each_key)\n",
    "        else:\n",
    "            dict2[chromosome] = [each_key]\n",
    "    return dict2\n",
    "\n",
    "def find_overlapping_ranges(position, position_dict, chromosome, how_close):\n",
    "    ranges_to_search = position_dict[chromosome]  # only look within the same chromosome\n",
    "    overlapping_ranges = []  # to store the ranges that overlap the query position\n",
    "    ranges_to_search_sorted = sorted(ranges_to_search)\n",
    "    x1 = position - how_close\n",
    "    x2 = position + how_close\n",
    "    for each_range in ranges_to_search_sorted:\n",
    "        range_split = each_range.split('.')\n",
    "        y1 = int(range_split[1])\n",
    "        y2 = int(range_split[2])\n",
    "        if x1 <= y2 and y1 <= x2:\n",
    "            overlapping_ranges.append(each_range)\n",
    "        if y1 > x2:  # beginning of search range is beyond end of query range, no point in searching past this\n",
    "            break\n",
    "    return overlapping_ranges\n",
    "\n",
    "def connect_back_to_genes(dict1, dict2, dict3):\n",
    "    for each_chr, positions in dict1.items():\n",
    "        for each_pos, data in positions.items():\n",
    "            genes = []\n",
    "            for each_range in data:\n",
    "                connected_gene = dict2[each_range]\n",
    "                genes.append(connected_gene)\n",
    "            new_key = f'{each_chr}_{each_pos}'\n",
    "            flat_genes = [item for sublist in genes for item in sublist]\n",
    "            flat_genes = list(set(flat_genes))   # added to output only unique gene lists\n",
    "            gene_string = \",\".join(flat_genes)\n",
    "            dict3[new_key] = gene_string\n",
    "    return dict3\n",
    "\n",
    "def filter_df(df, col, filter_value):\n",
    "    filtered = df.loc[df[col] == filter_value]\n",
    "    return filtered\n",
    "\n",
    "p_o = read_in_excel(HiC)  # read in gene:interactingfragment:tissue df\n",
    "esc_p_o = filter_df(p_o, 'Tissue_type', 'Human ES cells')\n",
    "npc_p_o = filter_df(p_o, 'Tissue_type', 'Neural Progenitor Cell')\n",
    "msc_p_o = filter_df(p_o, 'Tissue_type', 'Mesenchymal Stem Cell')\n",
    "\n",
    "allcols = []\n",
    "for colname, d in {'genes': p_o, 'esc_genes': esc_p_o, 'npc_genes': npc_p_o, 'msc_genes': msc_p_o}.items():\n",
    "    p_o_dict = rows_to_dict(d, {})  # dict where keys are chr.start.end and value is a list of promoters interacting with it\n",
    "    p_o_pos_dict = create_range_dict(p_o_dict, {})    # dict of chr: chr.start.end of ranges\n",
    "\n",
    "    chr_grouped = df_vars.groupby('chromosome')\n",
    "    chr_output = {}\n",
    "    for each_chromosome, variants in chr_grouped:\n",
    "        positions = variants['pos_hg19'].tolist()\n",
    "        chr_output[each_chromosome] = {x: find_overlapping_ranges(int(x), p_o_pos_dict, each_chromosome, dist_from_var_HiC) for x in positions}   # pos: [chr.start.end] from ranges\n",
    "\n",
    "    to_genes = connect_back_to_genes(chr_output, p_o_dict, {})\n",
    "    print(len(to_genes.values()))\n",
    "    print(sum(len(x)!=0 for x in to_genes.values()))\n",
    "    output_df = pd.DataFrame.from_dict(to_genes, orient='index')\n",
    "    output_df.columns = [colname]\n",
    "    allcols.append(output_df)\n",
    "\n",
    "# join all output dfs\n",
    "output_df_HiC = pd.concat(allcols, axis=1)\n",
    "#print(output_df_HiC[output_df_HiC['genes'] != ''])\n",
    "\n",
    "\n",
    "# flatten the lists of multiple genes for each variant\n",
    "\n",
    "output_df_HiC['oligo'] = df_vars.index.tolist()\n",
    "output_df_HiC = output_df_HiC.set_index('oligo')\n",
    "output_df_HiC = output_df_HiC.replace('', 0)\n",
    "\n",
    "def clean_hic_lists(r, column):\n",
    "    genes = r[column]\n",
    "    if not genes == 0:\n",
    "        genes_l = genes.split(';')\n",
    "        clean_gs = ','.join(genes_l)\n",
    "        return clean_gs\n",
    "    else: \n",
    "        return 0\n",
    "\n",
    "output_df_HiC['genes'] = output_df_HiC.apply(lambda row: clean_hic_lists(row, 'genes'), axis=1)\n",
    "output_df_HiC['esc_genes'] = output_df_HiC.apply(lambda row: clean_hic_lists(row, 'esc_genes'), axis=1)\n",
    "output_df_HiC['msc_genes'] = output_df_HiC.apply(lambda row: clean_hic_lists(row, 'msc_genes'), axis=1)\n",
    "output_df_HiC['npc_genes'] = output_df_HiC.apply(lambda row: clean_hic_lists(row, 'npc_genes'), axis=1)\n",
    "#print(output_df_HiC[output_df_HiC['esc_genes'] != 0].tail(5))\n",
    "\n",
    "# convert each column to ensID\n",
    "output_df_HiC['genes_ens'] = output_df_HiC.apply(lambda row: convert(row, 'hgnc', 'genes'), axis=1)\n",
    "output_df_HiC['esc_genes_ens'] = output_df_HiC.apply(lambda row: convert(row, 'hgnc', 'esc_genes'), axis=1)\n",
    "output_df_HiC['msc_genes_ens'] = output_df_HiC.apply(lambda row: convert(row, 'hgnc', 'msc_genes'), axis=1)\n",
    "output_df_HiC['npc_genes_ens'] = output_df_HiC.apply(lambda row: convert(row, 'hgnc', 'npc_genes'), axis=1)\n",
    "\n",
    "# check\n",
    "print(output_df_HiC[output_df_HiC['esc_genes_ens'] != 0].tail(2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4688\n"
     ]
    }
   ],
   "source": [
    "print(len(output_df_HiC[output_df_HiC['genes_ens'] != 0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                 GH_genes_ns  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]       VCAM1,lnc-VCAM1-1,piR-33432-028,piR-53029-003   \n",
      "libA_seq255928_[chr1:101157210-101157409]  piR-59928-010,piR-53029-003,VCAM1,piR-33432-02...   \n",
      "\n",
      "                                              GH_genes_s  \n",
      "libA_seq36960_[chr1:101150064-101150263]   piR-33432-028  \n",
      "libA_seq255928_[chr1:101157210-101157409]  piR-33432-028  \n"
     ]
    }
   ],
   "source": [
    "## Genehancer data \n",
    "\n",
    "dist_fr_var_GH = 10000 # want to look for enhancer-gene connections in GeneHancer within 10kb of our variant (10kb on either side)\n",
    "\n",
    "def find_overlapping_ranges(position, position_dict, chromosome, how_close):    # position dict: {'chr': 'start.end'}\n",
    "    info = position_dict[chromosome]  # only look within the same chromosome\n",
    "    overlapping_ranges = []  # to store the ranges that overlap the query position\n",
    "    ranges_to_search = list(info)  # list of positions to search through\n",
    "    ranges_to_search_sorted = sorted(ranges_to_search)\n",
    "    x1 = position - how_close\n",
    "    x2 = position + how_close\n",
    "    for each_range in ranges_to_search_sorted:\n",
    "        range_split = each_range.split('.')\n",
    "        y1 = int(range_split[1])\n",
    "        y2 = int(range_split[2])\n",
    "        if x1 <= y2 and y1 <= x2:\n",
    "            overlapping_ranges.append(info[each_range])  # append the GHID\n",
    "        if y1 > x2:  # beginning of search range is beyond end of query range, no point in searching past this\n",
    "            break\n",
    "    return overlapping_ranges\n",
    "\n",
    "e_df = pd.read_csv(GH_enhancers, sep='\\t', header=0, index_col='GHid')  # enhancers, their locations, and GHids\n",
    "e_df['GHid'] = e_df.index.tolist()\n",
    "e_df['chr'] = 'chr' + e_df['chr'].astype(str)\n",
    "e_df['chrom_start_end'] = e_df['chr'].astype(str) + '.' + e_df['element_start'].astype(str) + '.' + e_df['element_end'].astype(str)\n",
    "#print(e_df.head())\n",
    "\n",
    "vg_df = pd.read_csv(GH_connections, sep='\\t', header=0, index_col=False)  # enhancer-gene connections, the gene and whether connection is elite or not\n",
    "vg_df['is_enhancer_elite'] = vg_df['GHid'].map(e_df['is_elite'])   # add col to enhancer-gene connection table which says whether the enhancer itself is elite or not\n",
    "vg_df['chr'] = vg_df['GHid'].map(e_df['chr'])\n",
    "vg_df['e_start'] = vg_df['GHid'].map(e_df['element_start'])\n",
    "vg_df['e_end'] = vg_df['GHid'].map(e_df['element_end'])\n",
    "vg_df['chrom_start_end'] = vg_df['chr'].astype(str) + '.' + vg_df['e_start'].astype(str) + '.' + vg_df['e_end'].astype(str)\n",
    "\n",
    "vg_df.set_index(['GHid', 'symbol'], inplace=True)\n",
    "vg_df.sort_index(inplace=True)\n",
    "vg_df.drop(columns=['combined_score'], inplace=True)\n",
    "\n",
    "# to use with stringent list\n",
    "filtered_vg_df = vg_df[(vg_df['is_elite'] == 1)]   # only filtering on v-g connections\n",
    "\n",
    "\n",
    "pos_dict = {}   # chromosome: {chr.start.end: GHID}\n",
    "chrom_grouped = e_df.groupby('chr')\n",
    "for each_chr, grp in chrom_grouped:\n",
    "    cs = dict(zip(grp['chrom_start_end'].tolist(), grp['GHid'].tolist()))   # {chrom.start.end: GHID}\n",
    "    pos_dict[each_chr] = cs\n",
    "\n",
    "chrom_grouped = df_vars.groupby('chromosome')  # find overlapping ranges in our variants by chromosome\n",
    "chr_output = {}\n",
    "for each_chr, variants in chrom_grouped:\n",
    "    positions = variants['pos_hg38'].tolist()\n",
    "    positions = [x for x in positions if not np.isnan(x)]\n",
    "    chr_output[each_chr] = {x: find_overlapping_ranges(int(x), pos_dict, each_chr, dist_fr_var_GH) for x in positions if x != 'None'}   # pos: [GHID] from ranges\n",
    "\n",
    "to_genes_ns = {}\n",
    "to_genes_s = {}\n",
    "\n",
    "for each_chromosome, positions in chr_output.items():\n",
    "    for each_position, connections in positions.items():\n",
    "        x_ns = []\n",
    "        x_s = []\n",
    "        for each_connection in connections:   # each_connection is the GHID\n",
    "            if each_connection in filtered_vg_df.index:\n",
    "                connected_genes_df_s = filtered_vg_df.loc[each_connection, :]\n",
    "                connected_genes_s = connected_genes_df_s.index.tolist()  # get the symbols from the other layer of multiindex\n",
    "                x_s.append(connected_genes_s)\n",
    "            if each_connection in vg_df.index:\n",
    "                connected_genes_df_ns = vg_df.loc[each_connection, :]\n",
    "                connected_genes_ns = connected_genes_df_ns.index.tolist()\n",
    "                x_ns.append(connected_genes_ns)\n",
    "            \n",
    "        a_ns = [gene for sublist in x_ns for gene in sublist if gene != '']\n",
    "        connected_genes_clean_ns = list(set(a_ns))\n",
    "        connected_genes_clean_string_ns = ','.join(connected_genes_clean_ns)\n",
    "        to_genes_ns[f'{each_chromosome}_{each_position}'] = connected_genes_clean_string_ns\n",
    "        \n",
    "        a_s = [gene for sublist in x_s for gene in sublist if gene != '']\n",
    "        connected_genes_clean_s = list(set(a_s))\n",
    "        connected_genes_clean_string_s = ','.join(connected_genes_clean_s)\n",
    "        to_genes_s[f'{each_chromosome}_{each_position}'] = connected_genes_clean_string_s\n",
    "        \n",
    "output_df_GH = pd.DataFrame(index=df_vars.index.tolist())        \n",
    "output_df_GH['GH_genes_ns'] = df_vars['chr_pos38'].map(to_genes_ns)\n",
    "output_df_GH['GH_genes_s'] = df_vars['chr_pos38'].map(to_genes_s)\n",
    "\n",
    "output_df_GH['GH_genes_ns'].fillna(0, inplace=True)\n",
    "output_df_GH['GH_genes_ns'].replace('',0, inplace=True)\n",
    "output_df_GH['GH_genes_s'].fillna(0, inplace=True)\n",
    "output_df_GH['GH_genes_s'].replace('', 0, inplace=True)\n",
    "\n",
    "print(output_df_GH[output_df_GH['GH_genes_ns'] != 0].head(2))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5442\n"
     ]
    }
   ],
   "source": [
    "print(len(output_df_GH[output_df_GH['GH_genes_ns'] != 0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                 GH_genes_ns  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]       VCAM1,lnc-VCAM1-1,piR-33432-028,piR-53029-003   \n",
      "libA_seq255928_[chr1:101157210-101157409]  piR-59928-010,piR-53029-003,VCAM1,piR-33432-02...   \n",
      "libA_seq20072_[chr1:101196756-101196955]           HSALNG0005548,ENSG00000228086,VCAM1,GPR88   \n",
      "libB_seq201217_[chr1:104059849-104060048]  AMY2B,RF00100-006,RN7SKP285,LOC101928436,piR-3...   \n",
      "libA_seq100098_[chr1:104059992-104060191]  AMY2B,RF00100-006,RN7SKP285,LOC101928436,piR-3...   \n",
      "\n",
      "                                                   GH_genes_s  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]        piR-33432-028   \n",
      "libA_seq255928_[chr1:101157210-101157409]       piR-33432-028   \n",
      "libA_seq20072_[chr1:101196756-101196955]                VCAM1   \n",
      "libB_seq201217_[chr1:104059849-104060048]  LOC101928436,RNPC3   \n",
      "libA_seq100098_[chr1:104059992-104060191]  LOC101928436,RNPC3   \n",
      "\n",
      "                                          GH_genes_ns_ens_split  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]                      0   \n",
      "libA_seq255928_[chr1:101157210-101157409]                     0   \n",
      "libA_seq20072_[chr1:101196756-101196955]        ENSG00000228086   \n",
      "libB_seq201217_[chr1:104059849-104060048]                     0   \n",
      "libA_seq100098_[chr1:104059992-104060191]                     0   \n",
      "\n",
      "                                                                      GH_genes_ns_hgnc_split  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]       VCAM1,lnc-VCAM1-1,piR-33432-028,piR-53029-003   \n",
      "libA_seq255928_[chr1:101157210-101157409]  piR-59928-010,piR-53029-003,VCAM1,piR-33432-02...   \n",
      "libA_seq20072_[chr1:101196756-101196955]                           HSALNG0005548,VCAM1,GPR88   \n",
      "libB_seq201217_[chr1:104059849-104060048]  AMY2B,RF00100-006,RN7SKP285,LOC101928436,piR-3...   \n",
      "libA_seq100098_[chr1:104059992-104060191]  AMY2B,RF00100-006,RN7SKP285,LOC101928436,piR-3...   \n",
      "\n",
      "                                          GH_genes_s_ens_split  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]                     0   \n",
      "libA_seq255928_[chr1:101157210-101157409]                    0   \n",
      "libA_seq20072_[chr1:101196756-101196955]                     0   \n",
      "libB_seq201217_[chr1:104059849-104060048]                    0   \n",
      "libA_seq100098_[chr1:104059992-104060191]                    0   \n",
      "\n",
      "                                          GH_genes_s_hgnc_split  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]          piR-33432-028   \n",
      "libA_seq255928_[chr1:101157210-101157409]         piR-33432-028   \n",
      "libA_seq20072_[chr1:101196756-101196955]                  VCAM1   \n",
      "libB_seq201217_[chr1:104059849-104060048]    LOC101928436,RNPC3   \n",
      "libA_seq100098_[chr1:104059992-104060191]    LOC101928436,RNPC3   \n",
      "\n",
      "                                          GH_genes_ns_ens_split_conv  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]                           0   \n",
      "libA_seq255928_[chr1:101157210-101157409]                          0   \n",
      "libA_seq20072_[chr1:101196756-101196955]                           0   \n",
      "libB_seq201217_[chr1:104059849-104060048]                          0   \n",
      "libA_seq100098_[chr1:104059992-104060191]                          0   \n",
      "\n",
      "                                                                 GH_genes_ns_hgnc_split_conv  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]                                     ENSG00000162692   \n",
      "libA_seq255928_[chr1:101157210-101157409]                                    ENSG00000162692   \n",
      "libA_seq20072_[chr1:101196756-101196955]                     ENSG00000181656,ENSG00000162692   \n",
      "libB_seq201217_[chr1:104059849-104060048]  ENSG00000060718,ENSG00000222069,ENSG0000018594...   \n",
      "libA_seq100098_[chr1:104059992-104060191]  ENSG00000060718,ENSG00000222069,ENSG0000018594...   \n",
      "\n",
      "                                          GH_genes_s_ens_split_conv  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]                          0   \n",
      "libA_seq255928_[chr1:101157210-101157409]                         0   \n",
      "libA_seq20072_[chr1:101196756-101196955]                          0   \n",
      "libB_seq201217_[chr1:104059849-104060048]                         0   \n",
      "libA_seq100098_[chr1:104059992-104060191]                         0   \n",
      "\n",
      "                                          GH_genes_s_hgnc_split_conv  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]                           0   \n",
      "libA_seq255928_[chr1:101157210-101157409]                          0   \n",
      "libA_seq20072_[chr1:101196756-101196955]             ENSG00000162692   \n",
      "libB_seq201217_[chr1:104059849-104060048]            ENSG00000185946   \n",
      "libA_seq100098_[chr1:104059992-104060191]            ENSG00000185946   \n",
      "\n",
      "                                                                            GH_genes_ns_HGNC  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]       VCAM1,lnc-VCAM1-1,piR-33432-028,piR-53029-003   \n",
      "libA_seq255928_[chr1:101157210-101157409]  piR-59928-010,piR-53029-003,VCAM1,piR-33432-02...   \n",
      "libA_seq20072_[chr1:101196756-101196955]                           HSALNG0005548,VCAM1,GPR88   \n",
      "libB_seq201217_[chr1:104059849-104060048]  AMY2B,RF00100-006,RN7SKP285,LOC101928436,piR-3...   \n",
      "libA_seq100098_[chr1:104059992-104060191]  AMY2B,RF00100-006,RN7SKP285,LOC101928436,piR-3...   \n",
      "\n",
      "                                              GH_genes_s_HGNC  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]        piR-33432-028   \n",
      "libA_seq255928_[chr1:101157210-101157409]       piR-33432-028   \n",
      "libA_seq20072_[chr1:101196756-101196955]                VCAM1   \n",
      "libB_seq201217_[chr1:104059849-104060048]  LOC101928436,RNPC3   \n",
      "libA_seq100098_[chr1:104059992-104060191]  LOC101928436,RNPC3   \n",
      "\n",
      "                                                                             GH_genes_ns_ENS  \\\n",
      "libA_seq36960_[chr1:101150064-101150263]                                     ENSG00000162692   \n",
      "libA_seq255928_[chr1:101157210-101157409]                                    ENSG00000162692   \n",
      "libA_seq20072_[chr1:101196756-101196955]     ENSG00000228086,ENSG00000181656,ENSG00000162692   \n",
      "libB_seq201217_[chr1:104059849-104060048]  ENSG00000060718,ENSG00000222069,ENSG0000018594...   \n",
      "libA_seq100098_[chr1:104059992-104060191]  ENSG00000060718,ENSG00000222069,ENSG0000018594...   \n",
      "\n",
      "                                            GH_genes_s_ENS  \n",
      "libA_seq36960_[chr1:101150064-101150263]                 0  \n",
      "libA_seq255928_[chr1:101157210-101157409]                0  \n",
      "libA_seq20072_[chr1:101196756-101196955]   ENSG00000162692  \n",
      "libB_seq201217_[chr1:104059849-104060048]  ENSG00000185946  \n",
      "libA_seq100098_[chr1:104059992-104060191]  ENSG00000185946  \n"
     ]
    }
   ],
   "source": [
    "# GH genes are a mix of ENSG and HGNC\n",
    "# iterate through each gene and either convert or keep to new list\n",
    "# first split into two columns\n",
    "\n",
    "def split_ens_and_hgnc(r, col):\n",
    "    genes = r[col]\n",
    "    if not genes == 0:   \n",
    "        genes_l = genes.split(',')\n",
    "        ens_l = []\n",
    "        hgnc_l = []\n",
    "        for g in genes_l:\n",
    "            if g.startswith('ENSG'):\n",
    "                ens_l.append(g)\n",
    "            else:\n",
    "                hgnc_l.append(g)\n",
    "        if len(ens_l) > 0:\n",
    "            ens = ','.join(ens_l)\n",
    "        if len(ens_l) == 0:\n",
    "            ens = 0\n",
    "        if len(hgnc_l) > 0:\n",
    "            hgnc = ','.join(hgnc_l)\n",
    "        if len(hgnc_l) == 0:\n",
    "            hgnc = 0\n",
    "        return ens, hgnc\n",
    "    else:\n",
    "        return 0, 0\n",
    "\n",
    "output_df_GH[['GH_genes_ns_ens_split', 'GH_genes_ns_hgnc_split']] = \\\n",
    "    output_df_GH.apply(lambda row: pd.Series(split_ens_and_hgnc(row, 'GH_genes_ns')), axis=1)\n",
    "\n",
    "output_df_GH[['GH_genes_s_ens_split', 'GH_genes_s_hgnc_split']] = \\\n",
    "    output_df_GH.apply(lambda row: pd.Series(split_ens_and_hgnc(row, 'GH_genes_s')), axis=1)\n",
    "\n",
    "#print(output_df_GH[output_df_GH['GH_genes_ns'] == 0].head(5))\n",
    "\n",
    "output_df_GH['GH_genes_ns_ens_split_conv'] = output_df_GH.apply(lambda row: convert(row, 'ens', 'GH_genes_ns_ens_split'), axis=1)\n",
    "output_df_GH['GH_genes_ns_hgnc_split_conv'] = output_df_GH.apply(lambda row: convert(row, 'hgnc', 'GH_genes_ns_hgnc_split'), axis=1)\n",
    "output_df_GH['GH_genes_s_ens_split_conv'] = output_df_GH.apply(lambda row: convert(row, 'ens', 'GH_genes_s_ens_split'), axis=1)\n",
    "output_df_GH['GH_genes_s_hgnc_split_conv'] = output_df_GH.apply(lambda row: convert(row, 'hgnc', 'GH_genes_s_hgnc_split'), axis=1)\n",
    "\n",
    "def merge_with_conv_col(r, col1, col2):\n",
    "    g1 = r[col1]\n",
    "    g2 = r[col2]\n",
    "    if g1 != 0:\n",
    "        g1_l = g1.split(',')\n",
    "    if g1 == 0:\n",
    "        g1_l = []\n",
    "    if g2 != 0:\n",
    "        g2_l = g2.split(',')\n",
    "    if g2 == 0:\n",
    "        g2_l = []\n",
    "    g_l = g1_l + g2_l\n",
    "    merged_s = ','.join(g_l)\n",
    "    if len(merged_s) == 0:\n",
    "        merged_s = 0\n",
    "    return merged_s\n",
    "\n",
    "output_df_GH['GH_genes_ns_HGNC'] = \\\n",
    "    output_df_GH.apply(lambda row: merge_with_conv_col(row, 'GH_genes_ns_hgnc_split', 'GH_genes_ns_ens_split_conv'), axis=1)\n",
    "\n",
    "output_df_GH['GH_genes_s_HGNC'] = \\\n",
    "    output_df_GH.apply(lambda row: merge_with_conv_col(row, 'GH_genes_s_hgnc_split', 'GH_genes_s_ens_split_conv'), axis=1)\n",
    "\n",
    "output_df_GH['GH_genes_ns_ENS'] = \\\n",
    "    output_df_GH.apply(lambda row: merge_with_conv_col(row, 'GH_genes_ns_ens_split', 'GH_genes_ns_hgnc_split_conv'), axis=1)\n",
    "\n",
    "output_df_GH['GH_genes_s_ENS'] = \\\n",
    "    output_df_GH.apply(lambda row: merge_with_conv_col(row, 'GH_genes_s_ens_split', 'GH_genes_s_hgnc_split_conv'), axis=1)\n",
    "\n",
    "\n",
    "print(output_df_GH[output_df_GH['GH_genes_ns_ENS'] != 0].head(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5017\n"
     ]
    }
   ],
   "source": [
    "print(len(output_df_GH[output_df_GH['GH_genes_ns_ENS'] != 0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                          eQTL_genes_ENS  \\\n",
      "libB_seq57029_[chr1:1136106-1136305]   ENSG00000175756,ENSG00000225905,ENSG0000016257...   \n",
      "libA_seq229534_[chr1:1197948-1198147]  ENSG00000225905,ENSG00000162572,ENSG0000016007...   \n",
      "libB_seq264324_[chr1:1203069-1203268]  ENSG00000175756,ENSG00000225905,ENSG0000016257...   \n",
      "libB_seq75291_[chr1:1500965-1501164]   ENSG00000235098,ENSG00000215915,ENSG0000020509...   \n",
      "libA_seq35918_[chr1:1839919-1840118]   ENSG00000235098,ENSG00000231050,ENSG0000024833...   \n",
      "\n",
      "                                                                         eQTL_genes_HGNC  \n",
      "libB_seq57029_[chr1:1136106-1136305]   TTLL10,RNF223,ANKRD65,INTS11,ACAP3,TTLL10-AS1,...  \n",
      "libA_seq229534_[chr1:1197948-1198147]  CPTP,ANKRD65,INTS11,MXRA8,PERM1,ACAP3,TTLL10-A...  \n",
      "libB_seq264324_[chr1:1203069-1203268]  CPTP,ANKRD65,INTS11,MXRA8,ACAP3,MRPL20-DT,SCNN...  \n",
      "libB_seq75291_[chr1:1500965-1501164]   TTLL10,VWA1,ATAD3C,ANKRD65,SSU72,TMEM240,MRPL2...  \n",
      "libA_seq35918_[chr1:1839919-1840118]   SLC35E2A,SAMD11,ANKRD65,TMEM52,NADK,CDK11B,MMP...  \n"
     ]
    }
   ],
   "source": [
    "# eQTL data\n",
    "\n",
    "output_df_eqtl = pd.DataFrame(index=df_vars.index.tolist())\n",
    "\n",
    "def clean_eqtls(r):\n",
    "    eqtls = r['eQTL_dist_1000bp_v8_genes_unique']\n",
    "    if eqtls != 0:\n",
    "        eqtls_l = list(set(eqtls.split(',')))\n",
    "        eqtls_s = ','.join(eqtls_l)\n",
    "    if eqtls == 0:\n",
    "        eqtls_s = 0\n",
    "    return eqtls_s\n",
    "\n",
    "output_df_eqtl['eQTL_genes_ENS'] = df_vars.apply(lambda row: clean_eqtls(row), axis=1)\n",
    "output_df_eqtl['eQTL_genes_HGNC'] = output_df_eqtl.apply(lambda row: convert(row, 'ens', 'eQTL_genes_ENS'), axis=1)\n",
    "\n",
    "print(output_df_eqtl[output_df_eqtl['eQTL_genes_ENS'] != 0].head(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9503\n"
     ]
    }
   ],
   "source": [
    "print(len(output_df_eqtl[output_df_eqtl['eQTL_genes_ENS'] != 0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                              prom_ENS  prom_HGNC\n",
      "libB_seq51614_[chr1:1847036-1847235]   ENSG00000169885     CALML6\n",
      "libA_seq139784_[chr1:2355646-2355845]  ENSG00000149527      PLCH2\n",
      "libB_seq221362_[chr1:6244322-6244521]  ENSG00000116254       CHD5\n",
      "libB_seq249667_[chr1:8887208-8887407]  ENSG00000229716  RPL23AP19\n",
      "libA_seq305430_[chr1:9002429-9002628]  ENSG00000131686        CA6\n"
     ]
    }
   ],
   "source": [
    "# promoter data\n",
    "# for both ns and s lists: take all overlapping TSSs\n",
    "\n",
    "output_df_prom = pd.DataFrame(index=df_vars.index.tolist())\n",
    "\n",
    "df_vars['Overlapping_TSSs_Ensembl'] = df_vars['Overlapping_TSSs_Ensembl'].fillna(0)\n",
    "\n",
    "output_df_prom['prom_ENS'] = df_vars['Overlapping_TSSs_Ensembl']\n",
    "output_df_prom['prom_HGNC'] = output_df_prom.apply(lambda row: convert(row, 'ens', 'prom_ENS'), axis=1)\n",
    "\n",
    "print(output_df_prom[output_df_prom['prom_HGNC'] != 0].head(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1466\n"
     ]
    }
   ],
   "source": [
    "print(len(output_df_prom[output_df_prom['prom_ENS'] != 0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                       oligo  \\\n",
      "libB_seq57029_[chr1:1136106-1136305]    libB_seq57029_[chr1:1136106-1136305]   \n",
      "libA_seq229534_[chr1:1197948-1198147]  libA_seq229534_[chr1:1197948-1198147]   \n",
      "libB_seq264324_[chr1:1203069-1203268]  libB_seq264324_[chr1:1203069-1203268]   \n",
      "libB_seq75291_[chr1:1500965-1501164]    libB_seq75291_[chr1:1500965-1501164]   \n",
      "libA_seq35918_[chr1:1839919-1840118]    libA_seq35918_[chr1:1839919-1840118]   \n",
      "\n",
      "                                      promoter_HGNC promoter_ENS ns_HiC_HGNC  \\\n",
      "libB_seq57029_[chr1:1136106-1136305]              0            0           0   \n",
      "libA_seq229534_[chr1:1197948-1198147]             0            0           0   \n",
      "libB_seq264324_[chr1:1203069-1203268]             0            0           0   \n",
      "libB_seq75291_[chr1:1500965-1501164]              0            0           0   \n",
      "libA_seq35918_[chr1:1839919-1840118]              0            0           0   \n",
      "\n",
      "                                      s_ESC_HiC_HGNC s_Hob_HiC_HGNC  \\\n",
      "libB_seq57029_[chr1:1136106-1136305]               0              0   \n",
      "libA_seq229534_[chr1:1197948-1198147]              0              0   \n",
      "libB_seq264324_[chr1:1203069-1203268]              0              0   \n",
      "libB_seq75291_[chr1:1500965-1501164]               0              0   \n",
      "libA_seq35918_[chr1:1839919-1840118]               0              0   \n",
      "\n",
      "                                      s_NPC_HiC_HGNC ns_HiC_ENS s_ESC_HiC_ENS  \\\n",
      "libB_seq57029_[chr1:1136106-1136305]               0          0             0   \n",
      "libA_seq229534_[chr1:1197948-1198147]              0          0             0   \n",
      "libB_seq264324_[chr1:1203069-1203268]              0          0             0   \n",
      "libB_seq75291_[chr1:1500965-1501164]               0          0             0   \n",
      "libA_seq35918_[chr1:1839919-1840118]               0          0             0   \n",
      "\n",
      "                                      s_Hob_HiC_ENS s_NPC_HiC_ENS ns_GH_HGNC  \\\n",
      "libB_seq57029_[chr1:1136106-1136305]              0             0          0   \n",
      "libA_seq229534_[chr1:1197948-1198147]             0             0          0   \n",
      "libB_seq264324_[chr1:1203069-1203268]             0             0          0   \n",
      "libB_seq75291_[chr1:1500965-1501164]              0             0          0   \n",
      "libA_seq35918_[chr1:1839919-1840118]              0             0          0   \n",
      "\n",
      "                                      s_GH_HGNC ns_GH_ENS s_GH_ENS  \\\n",
      "libB_seq57029_[chr1:1136106-1136305]          0         0        0   \n",
      "libA_seq229534_[chr1:1197948-1198147]         0         0        0   \n",
      "libB_seq264324_[chr1:1203069-1203268]         0         0        0   \n",
      "libB_seq75291_[chr1:1500965-1501164]          0         0        0   \n",
      "libA_seq35918_[chr1:1839919-1840118]          0         0        0   \n",
      "\n",
      "                                                                               eQTL_HGNC  \\\n",
      "libB_seq57029_[chr1:1136106-1136305]   TTLL10,RNF223,ANKRD65,INTS11,ACAP3,TTLL10-AS1,...   \n",
      "libA_seq229534_[chr1:1197948-1198147]  CPTP,ANKRD65,INTS11,MXRA8,PERM1,ACAP3,TTLL10-A...   \n",
      "libB_seq264324_[chr1:1203069-1203268]  CPTP,ANKRD65,INTS11,MXRA8,ACAP3,MRPL20-DT,SCNN...   \n",
      "libB_seq75291_[chr1:1500965-1501164]   TTLL10,VWA1,ATAD3C,ANKRD65,SSU72,TMEM240,MRPL2...   \n",
      "libA_seq35918_[chr1:1839919-1840118]   SLC35E2A,SAMD11,ANKRD65,TMEM52,NADK,CDK11B,MMP...   \n",
      "\n",
      "                                                                                eQTL_ENS  \n",
      "libB_seq57029_[chr1:1136106-1136305]   ENSG00000175756,ENSG00000225905,ENSG0000016257...  \n",
      "libA_seq229534_[chr1:1197948-1198147]  ENSG00000225905,ENSG00000162572,ENSG0000016007...  \n",
      "libB_seq264324_[chr1:1203069-1203268]  ENSG00000175756,ENSG00000225905,ENSG0000016257...  \n",
      "libB_seq75291_[chr1:1500965-1501164]   ENSG00000235098,ENSG00000215915,ENSG0000020509...  \n",
      "libA_seq35918_[chr1:1839919-1840118]   ENSG00000235098,ENSG00000231050,ENSG0000024833...  \n"
     ]
    }
   ],
   "source": [
    "# final df for reporting gene lists, ns = not-stringent, s = stringent\n",
    "df2 = pd.DataFrame(index=df_vars.index.tolist())\n",
    "df2['oligo'] = df2.index.tolist()\n",
    "\n",
    "# promoter (same between ns and s)\n",
    "df2['promoter_HGNC'] = output_df_prom['prom_HGNC']\n",
    "df2['promoter_ENS'] = output_df_prom['prom_ENS']\n",
    "\n",
    "# HiC HGNC + ENSid\n",
    "df2['ns_HiC_HGNC'] = output_df_HiC['genes']\n",
    "df2['s_ESC_HiC_HGNC'] = output_df_HiC['esc_genes']\n",
    "df2['s_Hob_HiC_HGNC'] = output_df_HiC['msc_genes']\n",
    "df2['s_NPC_HiC_HGNC'] = output_df_HiC['npc_genes']\n",
    "\n",
    "df2['ns_HiC_ENS'] = output_df_HiC['genes_ens']\n",
    "df2['s_ESC_HiC_ENS'] = output_df_HiC['esc_genes_ens']\n",
    "df2['s_Hob_HiC_ENS'] = output_df_HiC['msc_genes_ens']\n",
    "df2['s_NPC_HiC_ENS'] = output_df_HiC['npc_genes_ens']\n",
    "\n",
    "# GH HGNC + ENSid (s is not cell-type specific)\n",
    "df2['ns_GH_HGNC'] = output_df_GH['GH_genes_ns_HGNC']\n",
    "df2['s_GH_HGNC'] = output_df_GH['GH_genes_s_HGNC']\n",
    "\n",
    "df2['ns_GH_ENS'] = output_df_GH['GH_genes_ns_ENS']\n",
    "df2['s_GH_ENS'] = output_df_GH['GH_genes_s_ENS']\n",
    "\n",
    "# eQTL, same between ns and s\n",
    "df2['eQTL_HGNC'] = output_df_eqtl['eQTL_genes_HGNC']\n",
    "df2['eQTL_ENS'] = output_df_eqtl['eQTL_genes_ENS']\n",
    "\n",
    "print(df2.head(5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine(r, columns):\n",
    "    genes = []\n",
    "    for c in columns:\n",
    "        g = r[c]\n",
    "        if g != 0:\n",
    "            g_l = g.split(',')\n",
    "            genes.extend(g_l)\n",
    "        if g == 0:\n",
    "            continue\n",
    "    genes_unique = list(set(genes))\n",
    "    if len(genes_unique) > 0:\n",
    "        combined = ','.join(genes_unique)\n",
    "    if len(genes_unique) == 0:\n",
    "        combined = 0\n",
    "    return combined\n",
    "\n",
    "df2['ns_final_HGNC'] = df2.apply(lambda row: combine(row, ['promoter_HGNC', 'ns_HiC_HGNC', 'ns_GH_HGNC', 'eQTL_HGNC']), axis=1)\n",
    "df2['s_final_ESC_HGNC'] = df2.apply(lambda row: combine(row, ['promoter_HGNC', 's_ESC_HiC_HGNC', 's_GH_HGNC', 'eQTL_HGNC']), axis=1)\n",
    "df2['s_final_Hob_HGNC'] = df2.apply(lambda row: combine(row, ['promoter_HGNC', 's_Hob_HiC_HGNC', 's_GH_HGNC', 'eQTL_HGNC']), axis=1)\n",
    "df2['s_final_NPC_HGNC'] = df2.apply(lambda row: combine(row, ['promoter_HGNC', 's_NPC_HiC_HGNC', 's_GH_HGNC', 'eQTL_HGNC']), axis=1)\n",
    "\n",
    "df2['ns_final_ENS'] = df2.apply(lambda row: combine(row, ['promoter_ENS', 'ns_HiC_ENS', 'ns_GH_ENS', 'eQTL_ENS']), axis=1)\n",
    "df2['s_final_ESC_ENS'] = df2.apply(lambda row: combine(row, ['promoter_ENS', 's_ESC_HiC_ENS', 's_GH_ENS', 'eQTL_ENS']), axis=1)\n",
    "df2['s_final_Hob_ENS'] = df2.apply(lambda row: combine(row, ['promoter_ENS', 's_Hob_HiC_ENS', 's_GH_ENS', 'eQTL_ENS']), axis=1)\n",
    "df2['s_final_NPC_ENS'] = df2.apply(lambda row: combine(row, ['promoter_ENS', 's_NPC_HiC_ENS', 's_GH_ENS', 'eQTL_ENS']), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11207\n",
      "11073\n"
     ]
    }
   ],
   "source": [
    "print(len(df2[df2['ns_final_ENS'] != 0]))\n",
    "print(len(df2[df2['ns_final_HGNC'] != 0]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.0\n",
      "6.431694476666369\n",
      "{1: 2447, 2: 1696, 3: 1254, 4: 977, 5: 813, 6: 599, 7: 481, 8: 404, 9: 314, 10: 251, 11: 216, 12: 176, 13: 162, 15: 151, 14: 135, 18: 125, 17: 94, 16: 93, 19: 86, 20: 68, 22: 65, 21: 62, 24: 57, 25: 55, 23: 51, 26: 49, 30: 40, 27: 35, 28: 31, 35: 21, 33: 20, 32: 17, 39: 15, 31: 14, 34: 12, 36: 10, 37: 10, 44: 9, 29: 9, 40: 8, 56: 6, 45: 6, 46: 5, 41: 5, 43: 4, 48: 4, 42: 4, 53: 4, 38: 4, 47: 4, 49: 3, 64: 2, 67: 2, 57: 2, 73: 2, 58: 2, 59: 2, 55: 2, 52: 2, 62: 2, 66: 1, 74: 1, 60: 1, 68: 1, 61: 1, 54: 1, 70: 1, 71: 1}\n"
     ]
    }
   ],
   "source": [
    "x = []\n",
    "def find_num_genes_linked(r):\n",
    "    genes = r['ns_final_ENS']\n",
    "    if genes == 0:\n",
    "        return\n",
    "    #print(genes)\n",
    "    genes_l = genes.split(',')\n",
    "    #print(genes_l)\n",
    "    num_genes = len(genes_l)\n",
    "    #print(num_genes)\n",
    "    x.append(num_genes)\n",
    "df2.apply(lambda row: find_num_genes_linked(row), axis=1)\n",
    "print(np.median(x))\n",
    "print(np.mean(x))\n",
    "print(pd.Series(x).value_counts().to_dict())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill in blank gene connections for ns list with closest TSS (even if not within promoter range)\n",
    "\n",
    "df_vars['ensembl_gene_id_TSS'].fillna(0, inplace=True)\n",
    "df_vars['hgnc_symbol_TSS'].fillna(0, inplace=True)\n",
    "def fill_in_blanks(r, col, typ):\n",
    "    genes = r[col]\n",
    "    oligo = r['oligo']\n",
    "    if genes != 0:\n",
    "        return genes\n",
    "    if genes == 0: # cases where no genes were connected\n",
    "        if typ == 'hgnc':\n",
    "            last_resort = df_vars.loc[oligo, 'hgnc_symbol_TSS']\n",
    "        if typ == 'ens':\n",
    "            last_resort = df_vars.loc[oligo, 'ensembl_gene_id_TSS']\n",
    "        if last_resort == 0:\n",
    "            return 0\n",
    "        if last_resort != 0:\n",
    "            return last_resort\n",
    "\n",
    "df2['ns_final_HGNC'] = df2.apply(lambda row: fill_in_blanks(row, 'ns_final_HGNC', 'hgnc'), axis=1)\n",
    "df2['ns_final_ENS'] = df2.apply(lambda row: fill_in_blanks(row, 'ns_final_ENS', 'ens'), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14037\n"
     ]
    }
   ],
   "source": [
    "print(len(df2[df2['ns_final_ENS'] != 0]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df2.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.to_csv(f'{p_d}/final_lists.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(df2.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
